### PR TITLE
Add CI workflow to test VTK master branch

### PR DESCRIPTION
**Pull request by banesullivan**
This uses VTK's GitLab package index to download and install a wheel for the latest VTK (master branch). To install, run:

```
pip install vtk --pre --no-cache --extra-index-url https://wheels.vtk.org
```

Specifically, this branch uses the OSMesa wheel to prevent the need to xvfb

This runs as a cron job on `main`, has a `workflow_dispatch` trigger to run on request, and has a branch naming pattern for `'maint/vtk-master*'` to be able to open branches/PRs that will verify things are working against VTK master.

### CODE REVIEW

**Comment by banesullivan** [comment]
vtkExplicitStructuredGrid DeepCopy is broken on master. Tracking here: https://gitlab.kitware.com/vtk/vtk/-/issues/18758

**Comment by banesullivan** [comment]
## Threshold API changes

```
AttributeError: 'vtkmodules.vtkFiltersCore.vtkThreshold' object has no attribute 'ThresholdBetween'
```

There is a MR to re-add this method: https://gitlab.kitware.com/vtk/vtk/-/merge_requests/9709

There is also a discussion about these changes here: https://discourse.vtk.org/t/unnecessary-vtk-api-change/9929

## Other errors

Two errors we may need to look into on our side:

```
FAILED tests/plotting/test_plotting.py::test_left_button_down - Failed: DID NOT RAISE <class 'ValueError'>
FAILED tests/plotting/test_plotting.py::test_plot_composite_poly_complex - Failed: DID NOT WARN. No warnings of type (<class 'numpy.ComplexWarning'>,) were emitted.
```

**Comment by banesullivan** [comment]
@akaszynski, do you recall why you added this check in #732:

https://github.com/pyvista/pyvista/blob/4bc2064e68b08e9f0374c3f7935d608f2db80d8a/pyvista/plotting/plotting.py#L1774-L1777

This is no longer failing for VTK master branch and thus the `left_button_down` test will be updated here.

File: pyvista/core/filters/data_set.py [hunk 1]
```diff
@@ -1100,7 +1100,10 @@ def threshold(
                 raise ValueError(
                     f'Value range must be length one for a float value or two for min/max; not ({value}).'
                 )
-            alg.ThresholdBetween(value[0], value[1])
+            # alg.ThresholdBetween(value[0], value[1])
+            alg.SetLowerThreshold(value[0])
+            alg.SetUpperThreshold(value[1])
+            alg.SetThresholdFunction(_vtk.vtkThreshold.THRESHOLD_BETWEEN)
```
Comment (banesullivan) [inline pyvista/core/filters/data_set.py:1106]: May want to put this in a separate PR to land it sooner

Comment (adeak) [reply pyvista/core/filters/data_set.py:1106]: Does this mean https://github.com/pyvista/pyvista/pull/2881 was an insufficient fix? Or did VTK change further? I can't figure it out, sorry.

Comment (banesullivan) [reply pyvista/core/filters/data_set.py:1106]: I think VTK changed further. This is a confusing one for sure

Comment (banesullivan) [reply pyvista/core/filters/data_set.py:1106]: This was handled in #3750

**Comment by akaszynski** [comment]
> @akaszynski, do you recall why you added this check in #732:
> 
> https://github.com/pyvista/pyvista/blob/4bc2064e68b08e9f0374c3f7935d608f2db80d8a/pyvista/plotting/plotting.py#L1774-L1777
> 
> This is no longer failing for VTK master branch and thus the `left_button_down` test will be updated here.

It's been a while, but I do recall genuine segfaults and having to work around them. I agree with your workaround.

File: .github/workflows/vtk-pre-test.yml [hunk 2]
```diff
@@ -0,0 +1,74 @@
+name: VTK Master Testing
+
+on:
+  # TODO: remove
+  push:
+    branches:
+      - maint/vtk-master-testing
```
Comment (banesullivan) [inline .github/workflows/vtk-pre-test.yml:7]:
```suggestion
```

File: .github/workflows/vtk-pre-test.yml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  SHELLOPTS: 'errexit:pipefail'
+
+jobs:
+  Linux:
+    name: Linux VTK Master Testing
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: [3.7, 3.8, 3.9, "3.10"]
```
Comment (banesullivan) [inline .github/workflows/vtk-pre-test.yml:27]:
```suggestion
        python-version: [3.7, 3.8, 3.9, "3.10", "3.11"]
```

File: .github/workflows/vtk-pre-test.yml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+      - name: Set up vtk
+        run: |
+          pip uninstall vtk --yes
+          pip install vtk --pre --no-cache --extra-index-url https://gitlab.kitware.com/api/v4/projects/13/packages/pypi/simple
+
+      - uses: awalsh128/cache-apt-pkgs-action@v1.1.3
+        with:
+          packages: libgl1-mesa-glx xvfb python-tk
+          version: 3.0
```
Comment (banesullivan) [inline .github/workflows/vtk-pre-test.yml:61]: Let's use the new OSMesa wheel variants when they get published later this/next week and remove libgl1-mesa-glx

File: .github/workflows/vtk-pre-test.yml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+      - name: Install Testing Requirements
+        run: pip install -r requirements_test.txt
+
+      - name: Software Report
+        run: |
+          xvfb-run python -c "import pyvista; print(pyvista.Report()); from pyvista import examples; print('User data path:', examples.USER_DATA_PATH)"
```
Comment (banesullivan) [inline .github/workflows/vtk-pre-test.yml:68]: Note comment above and remove `xvfb-run`

File: .github/workflows/vtk-pre-test.yml (hunk 4 above)
Comment (akaszynski) [reply .github/workflows/vtk-pre-test.yml:61]: While we're here, why do we install `python-tk`?

File: .github/workflows/vtk-pre-test.yml (hunk 5 above)
Comment (akaszynski) [reply .github/workflows/vtk-pre-test.yml:68]: If we no longer need `xvfb` (and probably not `python-tk`) we can remove all OS packages.

File: .github/workflows/vtk-pre-test.yml (hunk 4 above)
Comment (banesullivan) [reply .github/workflows/vtk-pre-test.yml:61]: No idea... had the same thought. This could be an artifice from qt?

Comment (akaszynski) [reply .github/workflows/vtk-pre-test.yml:61]: Potentially.

Easier to ask forgiveness. Let's remove it and see.

Comment (banesullivan) [reply .github/workflows/vtk-pre-test.yml:61]: I'll do this in a separate PR that cleans up the other CI jobs and leverages the OSMesa wheel for the 9.2.5 release

(Also upgrading jobs like docs to 3.11 where possible)

**Comment by banesullivan** [comment]
Okay! CI is finally running with the `vtk_osmesa` package and tests are not segfaulting!

Unfortunately we have tests failures with the main branch of VTK

**Comment by banesullivan** [comment]
@pyvista/developers, this is ready for review. IMO, we should merge this without addressing the testing failures, as the PR is to set up a CI routine that will make it easier to debug PyVista+VTK master issues.

After merging, we should follow up with individual PRs that address the testing failures here so that we can have targetted development for specific issues.

File: .github/workflows/vtk-pre-test.yml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      - name: Install pyvista
+        run: pip install . --no-deps
+
+      - name: Install Testing Requirements
+        run: pip install -r requirements_test.txt
+
+      - name: Set up VTK with OSMesa
+        run: |
+          pip uninstall vtk -y
+          pip install vtk_osmesa --pre --no-cache --extra-index-url https://gitlab.kitware.com/api/v4/projects/13/packages/pypi/simple
```
Comment (banesullivan) [inline .github/workflows/vtk-pre-test.yml:56]: @akaszynski, @MatthewFlamm, @AlejandroFernandezLuces: this is the issue I am referring to in https://github.com/pyvista/pytest-pyvista/pull/41

Comment (MatthewFlamm) [reply .github/workflows/vtk-pre-test.yml:56]: I think this is a separate issue from the issue you linked, but the impact is sometimes encountered in installing that package.  We have a dependency for `vtk` in pyvista, but apparently `vtk_osmesa` can be used without problem.  But, this means that `pyvista` still looks for `vtk`.  The problem is in the dependencies for `pyvista` itself, not in `pytest-pyvista` IMO.  I dont know if there is a solution for an 'either/or' dependency resolution in pip specification.

Comment (banesullivan) [reply .github/workflows/vtk-pre-test.yml:56]: Indeed the issue is wit `pyvista` having `vtk` in its hard requirements. This is something I think we will have a lot more friction changing. And so a simpler fix is to just remove `pyvista` from `pytest-pyvista`'s hard requirements to work around this.

**Comment by banesullivan** [comment]
@pyvista/developers, I recommend merging this PR as is to help track testing failures against VTK `master`. The workflow here is manually triggered, so will not cause failures on PRs.

File: .github/workflows/vtk-pre-test.yml [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      - name: Install pyvista
+        run: pip install . --no-deps
+
+      - name: Install Testing Requirements
+        run: pip install -r requirements_test.txt
+
+      - name: Set up VTK with OSMesa
+        run: |
+          pip uninstall vtk -y
+          pip install vtk_osmesa --pre --no-cache --extra-index-url https://wheels.vtk.org
```
Comment (larsoner) [inline .github/workflows/vtk-pre-test.yml:56]: It would be good to have one run that uses VTK-osmesa and another that uses Xvfb-and-standard-VTK. The failure modes could be different and the latter is a (much more?) widely used test case.

**Review by larsoner — COMMENTED** [review]
> The workflow here is manually triggered, so will not cause failures on PRs.

Works for me as long as someone sets up to be notified on failures and actually fixes them. In MNE we have had a run for a while now where we install all sorts of stuff either from `main`/`master` including PyVista(Qt) and other stuff from daily/weekly CDNs (NumPy, SciPy, pandas, as of a couple of days ago VTK, etc.), but we *do* run it on all PRs rather than on a schedule. It means that once a month or so some CI breaks in someone's PR for an unrelated reason, but these are easy to ignore during review ("merge anyway ignoring requirements", or just don't add to "Required" list!) but also annoying enough that we fix them. It also catches stuff like using a deprecated APIs which would be missed until after merge by a scheduled run. You might consider trying this approach and seeing how onerous it is, and resorting to a scheduled one only if necessary.

But a scheduled run is better than no run at all so +1 from me either way! Hopefully a follow-up PR soon will make this green 🙂

**Comment by akaszynski** [comment]
> Works for me as long as someone sets up to be notified on failures and actually fixes them

I'm of the opinion that it's better to be notified of failures early on and have this workflow (at least for one version of python) be always on.

However, I'm reminded of [alert fatigue](https://www.atlassian.com/incident-management/on-call/alert-fatigue#:~:text=Alert%20fatigue%E2%80%94also%20known%20as,the%20sheer%20number%20of%20alerts.) If this fails every so often, it's acceptable. If it's all the time and we can't correct it, we should have the option of disabling it.


Since it's failing right now, let's keep this manual for now and consider enabling it by default later.

**Review by akaszynski — APPROVED** [review]

File: .github/workflows/vtk-pre-test.yml (hunk 7 above)
Comment (banesullivan) [reply .github/workflows/vtk-pre-test.yml:56]: I think I will put that off for another PR -- due to current time constraints and because I want to use the new OSMesa wheel throughout our testing suite.

**Comment by banesullivan** [comment]
> ... You might consider trying this approach and seeing how onerous it is, and resorting to a scheduled one only if necessary.

Thanks for the excellent example. I'd prefer not to have flaky CI jobs on PRs -- it signals that some things are okay to fail and then we as a community are quick to ignore that failure and soon other failures because we're used to merging with a red X. "alert fatigue" as @akaszynski pointed out. I've seen this be a big problem on some projects that now I'm very averse to it.

> Since it's failing right now, let's keep this manual for now and consider enabling it by default later.

Indeed - this is failing right now, and I think we need to raise an issue or two upstream in VTK before these things can be addressed. I suspect many failures moving forward are going to be things are broken upstream and thus we don't have a lot of control here in PyVista -- a further reason I'd prefer this to be a manual job we intentionally run to report issues upstream at once.

One such (serious) issue was discovered in this PR (see https://gitlab.kitware.com/vtk/vtk/-/issues/18758). I need to dig into the logs here and report any current failures
